langchain4j/src/main/java/dev/langchain4j/service/DefaultAiServices.java

```diff
@@ -166,6 +172,38 @@ public Object invoke(Object proxy, Method method, Object[] args) throws Exceptio
                             return new AiServiceTokenStream(messages, contents, context, memoryId); // TODO moderation
                         }
 
+                        // Backup default tools
+                        List<ToolSpecification> defaultToolSpecifications = context.toolSpecifications;
+                        Map<String, ToolExecutor> defaultToolExecutors = context.toolExecutors;
+
+                        // Begin tools
+                        if (toolsFromArgs.isPresent()) {
+                            // Overriding default tools by argument
+                            context.toolSpecifications = new ArrayList<>();
+                            context.toolExecutors = new HashMap<>();
+                            Map<ToolSpecification, ToolExecutor> newTools = toolsFromArgs.get().getTools();
+                            newTools.forEach((toolSpecification, toolExecutor) -> {
+                                context.toolSpecifications.add(toolSpecification);
+                                context.toolExecutors.put(toolSpecification.name(), toolExecutor);
+                            });
+                        } else if (context.toolProvider != null) {
+                            // Use provider to decide matching tools
+                            ChatMemory chatMemory = context.hasChatMemory()
+                                    ? context.chatMemory(memoryId)
+                                    : null;
+                            context.toolSpecifications = new ArrayList<>();
+                            context.toolExecutors = new HashMap<>();
+                            ToolProviderRequest providerRequest = new ToolProviderRequest(chatMemory, userMessage);
+                            ToolProviderResult providerResult = context.toolProvider.provideTools(providerRequest);
+                            if (providerResult != null) {
+                                // Overriding default tools by provider tools
+                                providerResult.getTools().forEach((toolSpecification, toolExecutor) -> {
+                                    context.toolSpecifications.add(toolSpecification);
```

Comment:
This is not the expected behavior, tools returned in `ToolProviderResult` should be used only for the current AI Service invocation, they should not be cached in the context